crates/sui-types/src/temporary_store.rs

```diff
@@ -435,33 +435,33 @@ impl<S> TemporaryStore<S> {
     ) -> Result<(), ExecutionError> {
         let mut objects_to_update = vec![];
         // Also charge gas for mutating the gas object in advance.
-        let gas_object_size = gas_object.object_size_for_gas_metering();
-        gas_object.storage_rebate = gas_status.charge_storage_mutation(
-            gas_object_size,
-            gas_object_size,
-            gas_object.storage_rebate.into(),
-        )?;
+        gas_object.charge_storage_mutation_no_size_change(gas_status)?;
         objects_to_update.push((
             SingleTxContext::gas(sender),
             gas_object.clone(),
             WriteKind::Mutate,
         ));
 
         for (object_id, (ctx, object, write_kind)) in &mut self.written {
-            let (old_object_size, storage_rebate) = self
-                .input_objects
-                .get(object_id)
-                .map(|old| (old.object_size_for_gas_metering(), old.storage_rebate))
-                .unwrap_or((0, 0));
```

Comment:
This line seems wrong to me (see TODO in new code). If the written object is a dynamic field that already exists, there will be no corresponding input object, and we'll incorrectly get 0 for both `old_object_size` and `storage_rebate`. This will (I think) under-charging for the update computation cost, and double-charging for the dynamic field storage.

Reply:
@tnowacki is the expert, but I think you're right that using `input_objects` doesn't work here.  Two alternatives come to mind:

- When the store is converted to effects, we record all the objects that were modified and their previous versions, could we move this computation to the effects where this information exists?
- Augment the store so that it saves the initial sizes of all objects it loads (including the dynamic ones) -- doable, but seems kind of intrusive.

Reply:
Oh this is very wrong lol. Means we are overcharging for dynamic fields. 
We can't unwrap/expect as is done in the new code, as it will fail everytime for dynamic fields. 

I would suggest the following messy fix (messy in the sense the abstractions don't feel clean to me)
- add the size (or the contents) of the loaded object to `loaded_child_objects` returned from the object runtime
- add the `loaded_child_objects` to the temporary store

We wanted to do something similar to this eventually anyway for crash recovery work, so maybe it isn't that messy? Just add a field to storage for setting loaded child objects with `ObjectID => (Version, Size)`

Reply:
@tnowacki your solution makes sense to me given the current code. However:
- I think we would need `ObjectID => (Version, Size, storage_rebate)` (it's not possible to derive the rebate from the size alone because the storage gas price can change)
- This is orthogonal to both your proposed solution and this PR, but I'm not sure we actually need to store the old size. The only place this gets used is:
```
    // Computation cost of a mutation is charged based on the sum of the old and new size.
    // This is because to update an object in the store, we have to erase the old one and
    // write a new one.
    let cost = NumBytes::new((old_size + new_size) as u64)
            .mul(*INIT_SUI_COST_TABLE.object_mutation_per_byte_cost);
```

This seems doesn't make a lot of sense to me--I think we should just charge for the new size. There's no per-tx computational cost associated with "erasing" old objects, since we'll just drop a bunch of old objects at once when we prune. So the net effect of this is basically just charging you double.